### PR TITLE
Fixes / improvements for RegularShape

### LINKED ISSUE #12461: Circle's animation jitter
Recently we have updated OpenLayers from version 4.6.5 to 6.5.

In OpenLayers 4.6.5 we used `snapToPixel` option set to false for `ol.style.Circle` for "accurate" rendering mode during the animation to prevent noticable jitter. This phenomenon was also described in the documentation for the version 4.6.5:

> Using true allows for "sharp" rendering (no blur), while using false allows for "accurate" rendering. Note that accuracy is important if the circle's position is animated. Without it, the circle may jitter noticeably. Default value is true.

Unfortunately, this parameter has been removed in version 5.2.0, because it should be no longer needed according to the upgrade notes:

> The renderer now snaps to integer pixels when no interaction or animation is running to get crisp rendering. During interaction or animation, it does not snap to integer pixels to avoid jitter.

However, when we compare the same animation from version 4.6.5 to the animation from version 6.5 we can see the difference in rendering circle's animation which leads to the visible jitter (it is most visible at the end of the animation when the circle's diameter is the greatest). We prepared also a sample video with animation comparison:
https://www.youtube.com/watch?v=XGZdVtwdHMA

We are wondering if this change is related to some kind of a change in how snapping to pixels work right now? Is there anything we can do to make this animation smoother?

### PR DESCRIPTION
**Pull request by MoonE**
Fixes #12461

- Lazily initialize the canvases, and do not create some that were never needed.
- Canvas size was too small if the miter length is larger than the stroke width
- Canvas size was too large by one stroke width for round line join
- Reduce canvas size more for bevel line join
- bevel/miter line join missing on last/first point because path was not closed before stroking
- Implement high pixel ratio regular shape for immediate renderer
- It should always double the number of points when radius2 is provided.
- Don't require radius1 to be greater than radius2

### CODE REVIEW

**Comment by ahocevar** [comment]
Great work @MoonE! Let me know if you want me to look into anything specifically, or if you need help finishing this pull request.

File: src/ol/style/RegularShape.js [hunk 1]
```diff
@@ -220,24 +220,25 @@ class RegularShape extends ImageStyle {
 
   /**
    * Get the image icon.
-   * @param {number} pixelRatio Pixel ratio.
+   * @param {number} [pixelRatio=1] Pixel ratio.
    * @return {HTMLCanvasElement} Image or Canvas element.
    * @api
    */
   getImage(pixelRatio) {
-    if (!this.canvas_[pixelRatio || 1]) {
```
Comment (MoonE) [inline src/ol/style/RegularShape.js:228]: @ahocevar, Do you know if this function is ever called with an undefined / null / zero pixel ration?
I suspect it is not, so checking for a pixel ratio > 0 is unnecessary here?

Comment (ahocevar) [reply src/ol/style/RegularShape.js:228]: Correct, this is unnecessary. The check was introduced with #11277, but I don't see a need for it either.

**Comment by MoonE** [comment]
Setting the correct canvas size works, except with very pointy star, then it is still beveled. Maybe there is some hardcoded limit?
Though I don't think it is a real problem.

**Review by ahocevar — APPROVED** [review]
This looks excellent, @MoonE! Please check the failing rendering tests - if it's just pixel shifts caused by the fix, you can replace the reference images with those provided in the [rendering test artifacts](https://github.com/openlayers/openlayers/suites/3207853612/artifacts/74193001). Let me know if you need help with that. Also, please remove the test example. Then you can merge. Thanks!

**Review by ahocevar — CHANGES_REQUESTED** [review]
There is a problem when rendering circles in vector tiles. A square is drawn around the circle. See this enlarged image from the failing `layer-vectortile-simple` rendering test:

Untitled

**Comment by MoonE** [comment]
That is not the latest image. This was for testing, but it is already removed when I remove the example

**Comment by ahocevar** [comment]
Ah cool. While I'm at checking the rendering tests, I could add a commit with the updated reference images. Do you want me to do that, or are you already at it?

**Review by ahocevar — APPROVED** [review]
I just pushed that commit, @MoonE. Now rendering tests should pass. If they do, please merge.

**Comment by MoonE** [comment]
Thanks you, @ahocevar

**Pull request merged**